src/transactionIntentsFactories/smartContractTransactionIntentsFactory.spec.ts

```diff
@@ -0,0 +1,108 @@
+        const abiDeployIntent = abiSmartContractIntentsFactory.createTransactionIntentForDeploy(sender, adderByteCode.valueOf(), gasLimit, args);
+
+        assert.equal(deployIntent.sender, "erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th");
+        assert.equal(deployIntent.receiver, CONTRACT_DEPLOY_ADDRESS);
+        assert.isDefined(deployIntent.data);
+
+        if (deployIntent.data) {
+            const expectedGasLimit = 6000000 + 50000 + 1500 * deployIntent.data.length;
+            assert.equal(deployIntent.gasLimit.valueOf(), expectedGasLimit);
+        }
+        assert.equal(deployIntent.value, 0);
+
+        assert.deepEqual(deployIntent, abiDeployIntent);
+    });
+
+    it("should build execute intent", async function () {
+        const sender = Address.fromBech32("erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th");
+        const contract = Address.fromBech32("erd1qqqqqqqqqqqqqpgqhy6nl6zq07rnzry8uyh6rtyq0uzgtk3e69fqgtz9l4");
+        const func = "add";
+        const gasLimit = 6000000;
+        const args = [new U32Value(7)];
+
+        const deployIntent = smartContractIntentsFactory.createTransactionIntentForExecute(sender, contract, func, gasLimit, args);
+        const abiDeployIntent = abiSmartContractIntentsFactory.createTransactionIntentForExecute(sender, contract, func, gasLimit, args);
+
+        assert.equal(deployIntent.sender, "erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th");
+        assert.equal(deployIntent.receiver, "erd1qqqqqqqqqqqqqpgqhy6nl6zq07rnzry8uyh6rtyq0uzgtk3e69fqgtz9l4");
+
+        assert.isDefined(deployIntent.data);
+        let decoder = new TextDecoder();
+        assert.equal(decoder.decode(deployIntent.data), "add@07");
+
+        assert.equal(deployIntent.gasLimit.valueOf(), 6059000);
+        assert.equal(deployIntent.value, 0);
+
+        assert.deepEqual(deployIntent, abiDeployIntent);
+    });
+
+    it("should build upgrade intent", async function () {
+        const sender = Address.fromBech32("erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th");
+        const contract = Address.fromBech32("erd1qqqqqqqqqqqqqpgqhy6nl6zq07rnzry8uyh6rtyq0uzgtk3e69fqgtz9l4");
+        const gasLimit = 6000000;
+        const args = [new U32Value(0)];
+
+        const deployIntent = smartContractIntentsFactory.createTransactionIntentForUpgrade(sender, contract, adderByteCode.valueOf(), gasLimit, args);
+        const abiDeployIntent = abiSmartContractIntentsFactory.createTransactionIntentForUpgrade(sender, contract, adderByteCode.valueOf(), gasLimit, args);
+
+        assert.equal(deployIntent.sender, "erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th");
+        assert.equal(deployIntent.receiver, "erd1qqqqqqqqqqqqqpgqhy6nl6zq07rnzry8uyh6rtyq0uzgtk3e69fqgtz9l4");
+        assert.isDefined(deployIntent.data);
+
+        if (deployIntent.data) {
```

Comment:
Can drop condition if not necessary. Below, refactor to not use `TextDecoder` (as above).

Reply:
Not using `TextDecoder`. Condition still necessary.

Reply:
Also here, I think you can use `!` operator to avoid this check:
`assert(checkIfByteArrayStartsWith(deployIntent.data!, "upgradeContract@"));`
`const expectedGasLimit = 6000000 + 50000 + 1500 * deployIntent.data!.length;`

Reply:
Now using the `!` operator.